### PR TITLE
Adding DirectProgramming Monte Carlo Pi DPC++ sample

### PR DESCRIPTION
**Pull request by ethanhirsch**
# Description

Adding the MonteCarloPi code sample to the repository. It uses monte carlo method for estimation of pi, leveraging the new DPC++ reduction extension.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Tested on both target platforms.

- [X] Command Line
- [X] Visual Studio

### CODE REVIEW

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/License.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Copyright 2020 Intel Corporation
```
Comment (JoeOster) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/License.txt:1]: Remove the date

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+# `Monte Carlo Pi` Sample
```
Comment (JoeOster) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/README.md:1]: The title should not be highlighted,

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+  queue q(default_selector{}, dpc_common::exception_handler);
+  std::cout << "Running on "
+            << q.get_device().get_info<sycl::info::device::name>() << std::endl;
+
+  try {
+    // Set up buffers
+    buffer imgplot_buf((rgb*)image_plot,
+                       range(img_dimensions * img_dimensions));
+    buffer coords_buf((coordinate*)coords, range(size_n));
+    buffer total_buf((int*)(&total), range(1));
```
Comment (keryell) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:82]:
```suggestion
    sycl::buffer total_buf(&total, range(1));
```

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  // Set up sycl queue
+  queue q(default_selector{}, dpc_common::exception_handler);
+  std::cout << "Running on "
+            << q.get_device().get_info<sycl::info::device::name>() << std::endl;
+
+  try {
+    // Set up buffers
+    buffer imgplot_buf((rgb*)image_plot,
+                       range(img_dimensions * img_dimensions));
+    buffer coords_buf((coordinate*)coords, range(size_n));
```
Comment (keryell) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:81]:
```suggestion
    sycl::buffer coords_buf(coords, range(size_n));
```

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    std::cout << "ERROR: size_n must be greater than or equal to size_wg\n"
+              << std::endl;
+    exit(1);
+  }
+
+  // Initialize random seed
+  srand(time(NULL));
+
+  // Allocate memory for the output image
+  rgb* image_plot = (rgb*)calloc(img_dimensions * img_dimensions, sizeof(rgb));
```
Comment (keryell) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:146]: Bad coding style:
- spaghetti C
- no check for bad alloc
- memory leak if exception thrown

Use a `std::vector` instead :-)

**Review by keryell — CHANGES_REQUESTED** [review]
You can simplify the code.
When using cast, use C++ cast instead of C cast in C++ code.

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    // Perform Monte Carlo simulation and reduce results
+    q.submit([&](handler& h) {
+      // Set up accessors
+      auto imgplot_acc = imgplot_buf.get_access<access::mode::write>(h);
+      auto coords_acc = coords_buf.get_access<access::mode::read_write>(h);
+      auto total_acc = total_buf.get_access<access::mode::read_write>(h);
+
+      // Monte Carlo Procedure + Reduction
+      h.parallel_for(nd_range<1>(num_wg * size_wg, size_wg),
+                     sycl::intel::reduction(total_acc, 0, std::plus<int>()),
```
[the remark this answers is not in the document]
Comment (ethanhirsch) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:92]: Testing on beta08 I get compiler errors unless I specify sycl::intel::reduction

Shall I keep this as-is?

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
+#include "monte_carlo_pi.hpp"
+
+#include <math.h>
+#include <stdlib.h>
+#include <time.h>
+
+#include <CL/sycl.hpp>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (ethanhirsch) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:8]: clang-format enforces the specific ordering. Should I override it?

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+#include "dpc_common.hpp"
+
+#define STB_IMAGE_IMPLEMENTATION
+#include "stb/stb_image.h"
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include "stb/stb_image_write.h"
+
+using namespace sycl;
```
[the remark this answers is not in the document]
Comment (ethanhirsch) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:17]: Thanks for your feedback, we will consider your input as we work on updates for the next version of the coding guidelines we are using.

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
+#include "monte_carlo_pi.hpp"
```
Comment (JoeOster) [inline DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:1]: Need to clean up include list a bit

Comment (ethanhirsch) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:1]: Hey Joe, could you provide some clarification here?

After formatting with clang-format, this current include list is the result.

**Comment by JoeOster** [comment]
@keryell - Hey Ronan, did we resolve all your changes? We are waiting for your approval or next steps prior to merging

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:16]: Thanks. Usefulness of `using namespace sycl;` is very simple to understand, specially for a non programmer. Understanding why not using it requires some programming skills...

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp:90]: Yes, keep it as is while the standard percolate down to the implementation.

File: DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (keryell) [reply DirectProgramming/DPC++/MapReduce/MonteCarloPi/src/monte_carlo_pi.cpp]: Oh, I missed the point that you use both old C includes with C++ includes. :-(
You should use only C++ headers I guess...

**Review by keryell — COMMENTED** [review]
Thanks. This is simpler and looks cleaner.

**Comment by JoeOster** [comment]
> Thanks. This is simpler and looks cleaner.

We appreciate your assistance in these reviews, so thank you

**Comment by ethanhirsch** [comment]
> Thanks. This is simpler and looks cleaner.

And thank you! I will continue to implement your feedback, it is very much appreciated.

**Pull request merged**